### PR TITLE
Step towards i18n

### PR DESCRIPTION
**Pull request by PiOverFour**
This is a proposition to move to a simpler structure in the JSON files, in order to later be able to import the strings into such a platform as Weblate (they have free plans for open source developers, upon request). From the commit message:

> Instead of an object of the form:
> 
>     {
>         "source": "Cannot parse the value.",
>         "description": "",
>         "id": "Cannot parse the value.",
>         "text": "Nelze analyzovat hodnotu."
>     },
> 
> we have:
> 
>     "Cannot parse the value.": "Nelze analyzovat hodnotu.",
> 
> It is less redundant, easier to read, and most of all easier to
> import into a translation platform.
> 
> The Python scripts converts existing text files to this format.

The Python script was not yet applied. It does 3 things:
- convert the existing texts to the new format
- unescape some HTML code (from Google?)
- unescape utf-8 characters (the `ensure_ascii` options)

Please not that the `html.unescape()` function is from Python3.

### CODE REVIEW

**Comment by darbyjohnston** [comment]
I'm definitely in favor of any changes that can improve the translation workflow, thanks for looking into this!

One thing to consider is that the "source" and "id" strings are not always the same as in your example, the "id" in the code can be an arbitrary string while the "source" text is what is used for translation. For example the id "Play once tooltip" is replaced with "Play the file once and then stop playback" (or it's translation) from the .text files. If we use the "id" strings as the new keys for the JSON files it may not provide enough context for translation (especially machine translation).

One possibility would be to replace the "id" strings in the code with the "source" strings from the current .text files. Another possibility could be to keep the "id" strings in the code and make a new intermediate file that maps between them and the keys in the new JSON files.

**Comment by PiOverFour** [comment]
Ah, I see what you mean. I don’t know what the best practice is for this, and it depends on what translation solution is chosen. Right now, without a translation program, I think the first solution would be easiest: use the full English text as ID ("source"). It would work right away.
On Weblate, either solution may be used. IDs are called "context", and may be either the full string or a kind of variable name, e. g. `play_once_tooltip`. They may also not be used at all. Then an automatic correspondence between this context, the source language string, and the translated string is generated and displayed. Compare for example [this translation page](https://hosted.weblate.org/translate/osmand/main/es_US/?q=state:%3E=translated):
image
with [this one](https://hosted.weblate.org/translate/gnu-mailman/mailman/pt/?q=state:%3Ctranslated):
image
and [this one](https://hosted.weblate.org/translate/godot-engine/godot/af/?q=state:%3Ctranslated):
image

So in short, I think the simplest solution is to just copy the full strings to the code.

**Comment by darbyjohnston** [comment]
In the first translation page, is the English source text also entered with Weblate? Is the workflow like: the context strings are uploaded, the English source text is entered for each context string, and then translations are added? One nice thing about that is a non-developer could write/edit the source text without having to go into the code.

I tried a quick test replacing the DJV "id" strings with the "text" strings and thought of a couple more possible issues:
* Using the full text as the "id" could make string comparisons more expensive, though this is probably not a big deal since most of the text is only looked up when the language is changed.
* Some of the English text has utf-8 characters, for example the names in the about dialog. Is there any issue with having utf-8 in C++ source files?
* This one is a bit of a pain, some of the "id" strings are serialized in the settings file and the current parsing logic expects them to not have any spaces. Maybe on serialization/de-serialization spaces are replaced with a different character.

**Comment by PiOverFour** [comment]
> In the first translation page, is the English source text also entered with Weblate? Is the workflow like: the context strings are uploaded, the English source text is entered for each context string, and then translations are added?

No, the source language file is usually added directly in the git repository. I don’t know about this particular project, but it was either generated from the source code, or more probably added manually to the strings file. Upon push, Weblate detects a new string to translate, or a modified source string.

> One nice thing about that is a non-developer could write/edit the source text without having to go into the code.

As far as I understand, this is a separate [permisson](https://docs.weblate.org/en/weblate-3.10.3/admin/access.html). So maybe not the standard workflow, but doable. Anyway, even for users not granted this permission, there is a way to discuss source strings.

> Some of the English text has utf-8 characters, for example the names in the about dialog. Is there any issue with having utf-8 in C++ source files?

I’m afraid the only C++ code I’ve ever written is the one from this pull request :D I’d expect it to introduce building complications for at least some platforms, but [this](https://stackoverflow.com/questions/331690/using-unicode-in-c-source-code) says it’s possible. It kind of works in my environment anyway:
image
But then the translated strings aren’t recognised:
image
If it’s only people’s names, it’s not a problem since they shouldn’t be translated or otherwise modified in the first place.

I don’t know about the other issues, but I guess putting the English text into the source may not be the simplest solution after all! It’s easier initially because it doesn’t require much manual work, but there will be drawbacks.
Another possibility is to just sanitise IDs before putting them into the source. That way there is no (de)serialisation problem, the strings are still readable by translators, and we don’t need to go over each of them to make clean names.

At the moment, the strings for the English UI are read from the binary and not the text file, right? If so, and if the IDs are to be changed to be distinct from the texts, I think strings should always be loaded from the translation files, even for the English UI.

**Comment by PiOverFour** [comment]
Thought about it some more. There actually are compelling reasons to use unique descriptions as IDs. Some texts which are identical in English may not be in other languages, because of grammatical gender or other language features, or simply because a given English word may have varying translations depending on context. Right now there are collisions.
Also, I used the French UI a bit today and noticed it was still poorly translated (I should know, I corrected it myself :D) I see several reasons for that, besides my sloppiness:
- the text format is cluttered; this would be alleviated by my proposed structure or by the use of a translation program/service ;
- the context is not obvious. I now believe that the original string is insufficient, and that a good, unique, manually chosen context/ID string is the better way to go.

This would require a bit of work. I’m willing to do it if you think we should proceed with this idea. Well, a first pass at least. I will probably make mistakes since I don’t understand some features, such as the Frame Store.

**Comment by darbyjohnston** [comment]
> I’m afraid the only C++ code I’ve ever written is the one from this pull request :D

Cool! Welcome to the club. :)

> I don’t know about the other issues, but I guess putting the English text into the source may not be the simplest solution after all! It’s easier initially because it doesn’t require much manual work, but there will be drawbacks.

I liked your proposal for simplifying but it does seem like some of the complexity is necessary.

> I think strings should always be loaded from the translation files, even for the English UI.

Indeed this is the way it currently works, the strings are read from the .text files at startup including the English versions. After startup the .text files are monitored so that any changes are re-read and updated in the UI.

> There actually are compelling reasons to use unique descriptions as IDs. Some texts which are identical in English may not be in other languages, because of grammatical gender or other language features, or simply because a given English word may have varying translations depending on context. Right now there are collisions.

That's a really good point. Maybe another nice feature about unique IDs is that it should be obvious when they are missing their source text, since they will show up in the UI and look out of place.

> This would require a bit of work. I’m willing to do it if you think we should proceed with this idea. Well, a first pass at least. I will probably make mistakes since I don’t understand some features, such as the Frame Store.

That would be great, thanks! So it sounds like we are talking about:
* Specify a new unique ID format
* Replace the existing strings in the source code with new IDs
* Change the .text files to a format compatible with Weblate
* Design a workflow for uploading/downloading the .text files to Weblate (could just be documentation at first, though CI would be nice at some point)

There are quite a few strings (about 1000?) so I can definitely help with the search and replace.

Another thing to think about is currently some of the strings are assembled from pieces so that variables can be inserted, like:

`DJV_TEXT("The file ") << fileName << DJV_TEXT(" could not be opened")
`

This isn't great since it prevents the translator from re-ordering where "fileName" should appear in the sentence. Something like this would be better:

`Format(DJV_TEXT("The file {0} could not be opened")).arg(fileName)
`

Or maybe like this with an ID instead of the source text;

`Format(DJV_TEXT("file_open_error {0}")).arg(fileName)
`

> I don’t understand some features, such as the Frame Store.

Sorry about that, DJV really needs better documentation which is something I'm not very good at. The "Frame Store" allows you to compare different frames of an animation. You can copy a frame into a buffer (historically called the "store"), switch to a different frame, and then toggle the view between that frame and the one in the buffer. It's a bit old school, there are probably ways of improving the UX to make it more obvious.

**Comment by PiOverFour** [comment]
I’ve made some progress!
The new structure is somewhat hierarchical, in what I think is logical from a translation point of view. For instance, you have:
```
startup
startup_choose_language
...
menu_file
menu_file_open
menu_file_recent
...
settings
settings_general
settings_general_section_tooltips
settings_general_enable_tooltips
...
```
It is still WIP but I believe it’s already much clearer. I also came up with a convoluted process to update the strings, and applied it to my branch.

- From the old IDs, generate clean new IDs in snake_case
  - Put them in a old_id → new_id map
- Edit those new IDs manually
- For each ID file, generate new translation files, using the new ID and the old translation string (by looking up the ID in the old_id → new_id map),
- For each string, search and replace the ID in the source files

This is a good start, but there are some problems and remaining things to do:
- The biggest one, which I can’t solve myself, is that DJV segfaults on startup :D This is due to the setting IDs having changed. When DJV tries to load the config file, it can’t find the proper keys. I don’t know where to look in the code, are those keys hardcoded? Should they be updated to match the translation IDs?
- Some strings should be duplicated, but since they originally had the same ID, a single string is called from different places
- The strings file names should follow (at least loosely) the naming scheme, instead of the source file whence they came. So you might have `djv_menu.en.text`, `djv_error.en.text`, `djv_settings.en.text`, etc. Or, if that’s too fine, group them in one or several files hierarchically.
- As you pointed out, some strings are split. It would be best to do as you suggest `Format(DJV_TEXT("file_open_error {0}")).arg(fileName) `. I don’t know how that would look in the C++ code, but the translation should be straightforward.

> Design a workflow for uploading/downloading the .text files to Weblate (could just be documentation at first, though CI would be nice at some point)

I’m not too sure how that will work yet. I guess it’s up to you to make a request to Weblate for hosting since you’re the author and maintainer — except if you’re into self-hosting…
In Weblate, you set up a remote repo, either basic git or GitHub. It is well-explained [here](https://docs.weblate.org/en/weblate-3.10.3/vcs.html#git). You basically give Weblate access to the GitHub repo through HTTPS or SSH (with key pairs), and it can pull from there. Once changes are done, it can create pull requests or push directly into the repo, through your GitHub user, with an access token.

I have never set up CI, in Weblate or elsewhere, so I can’t help much here.

**Comment by darbyjohnston** [comment]
> The new structure is somewhat hierarchical, in what I think is logical from a translation point of view. 

Looks like a good start!

> The biggest one, which I can’t solve myself, is that DJV segfaults on startup :D This is due to the setting IDs having changed. When DJV tries to load the config file, it can’t find the proper keys. I don’t know where to look in the code, are those keys hardcoded? Should they be updated to match the translation IDs?

The settings don't have very good error checking yet, so it might be easiest to increment the settings version which resets everything to default; it's located at the top of this file:
https://github.com/darbyjohnston/DJV/blob/master/lib/djvUI/SettingsSystem.cpp

> Some strings should be duplicated, but since they originally had the same ID, a single string is called from different places

I think that would be tricky to do with a script, we can probably sort it out manually as a final QA pass. 

> The strings file names should follow (at least loosely) the naming scheme, instead of the source file whence they came. So you might have djv_menu.en.text, djv_error.en.text, djv_settings.en.text, etc. Or, if that’s too fine, group them in one or several files hierarchically.

The idea for the current naming scheme is that an application may not need all of the translations, which is why they are broken up by library/binary name. So for example the small "djv_ls" command-line utility only needs the translation files "djvCore.en.text" and "djv_ls.en.text", while the "djv" viewer app needs "djvCore.en.text", "djvAV.en.text", "djvUI.en.text", "djv.en.text", etc. I would be a bit worried about the startup cost of parsing all of the combined JSON.

Also if we add another UI application to DJV at some point, like maybe a lightweight compositing tool, it could be confusing if strings from multiple apps are mixed together and there could also be id clashes if each app has a "file_open_tooltip" that needs to be customized.

How about using your naming scheme, but dividing the text files into sub-directories based upon which library/binary they are for? Maybe like:
```
$ROOT/etc/Text/djvCore/error.en.text
$ROOT/etc/Text/djvUI/error.en.text
$ROOT/etc/Text/djvViewApp/error.en.text
etc.
```

> As you pointed out, some strings are split. It would be best to do as you suggest Format(DJV_TEXT("file_open_error {0}")).arg(fileName) . I don’t know how that would look in the C++ code, but the translation should be straightforward.

Let's do that as a separate pass and I can start writing a string formatting class in parallel with the work you are doing.

> I can write some documentation once the system is up and running.

Thanks, that would be very helpful, probably just a list of steps people could follow on the contributing page. I'll check into creating the Weblate account.

**Comment by PiOverFour** [comment]
> The settings don't have very good error checking yet, so it might be easiest to increment the settings version

This wasn’t enough to prevent the crashes.The errors come from the `djv::UI::Settings::Style` strings. They are hardcoded, like `ColorRole::Background`. For now I’ve reverted my edits for these strings, they’d better be fixed separately since some of them are duplicates.

> > Some strings should be duplicated
> 
> I think that would be tricky to do with a script, we can probably sort it out manually as a final QA pass.

I agree. Duplicates can be found by a script, but modifying them requires knowing what they mean in context. I pushed a script to detect duplicates, for the next step. Here’s an [example output](https://github.com/darbyjohnston/DJV/files/4176323/duplicates.txt) for this version.

> The idea for the current naming scheme is that an application may not need all of the translations, which is why they are broken up by library/binary name.

Ooh, I didn’t think of that. Thanks for the explanation.

> How about using your naming scheme, but dividing the text files into sub-directories based upon which library/binary they are for?

It’s probably overkill then, because no file is longer than ~250 lines — except djvCore.en.text but it’s mostly fish names. So the strings could simply be grouped in each file.

> Let's do that as a separate pass and I can start writing a string formatting class in parallel with the work you are doing.

All right then! I renamed all the strings I could and left some untouched: the ones that induced the crash, some duplicates, and some whose function I didn’t know (e.g. `8-Bit Conversion`, `HUD`, `ParentChanged`).

**Comment by darbyjohnston** [comment]
> This wasn’t enough to prevent the crashes.

I think I see the issue, as you mentioned I can fix the crash and replace the strings separately.

> Here’s an example output for this version.

That's more that I thought, but shouldn't be too bad to fix manually.

> It’s probably overkill then, because no file is longer than ~250 lines — except djvCore.en.text but it’s mostly fish names. So the strings could simply be grouped in each file.

Sounds good, if the files get too large over time we can always revisit splitting them up. I'm sure some of the machine translated fish names came out quite bizarre. :) I was thinking of removing them from translation but they are handy for testing text rendering, even if the translations don't make sense.

> All right then! I renamed all the strings I could and left some untouched

Great, and thanks again! This should really help the translation process. Should I go ahead and merge the changes and start the manual cleanup?

I created a class for formatting strings to replace the broken up text, hopefully it is fairly straightforward. A string like this:

"The file {0} cannot be opened. The error is: {1}"

Will have the special arguments "{0}" and "{1}" replaced in the code like:

`message = String::Format(_getText(DJV_TEXT("error_file_open"))).arg(fileName).arg(error);
`

The arguments can be re-ordered in the translated string if necessary, the number between the curly brackets is used to correspond to the index of the argument in the code.

I also submitted the DJV project to Weblate though the website says it may take up to a month to process the submission.

**Comment by PiOverFour** [comment]
> That's more that I thought, but shouldn't be too bad to fix manually.

I can still help with those, but I don’t know if we can work in parallel.

> I'm sure some of the machine translated fish names came out quite bizarre. :)

Most of them are either all right or untranslated, but some are interesting. The one I like best is the Inland Threespine Stickleback, translated to French "Épinoche à trois épines de l’intérieur". Translated back, it would be something like "Stickleback with three thorns from the inside". Ouch!

> I was thinking of removing them from translation but they are handy for testing text rendering, even if the translations don't make sense.

I don’t think they’re really a problem. They will stand out against the rest of the string anyway.

> Should I go ahead and merge the changes and start the manual cleanup?

Yes! The only thing I’m not sure of is if I should leave those temporary scripts and ID files (old_IDs and new_IDs directories), but they can be removed at a later stage.

> 
> I created a class for formatting strings to replace the broken up text, hopefully it is fairly straightforward. A string like this:
> 
> "The file {0} cannot be opened. The error is: {1}"

It looks good, much easier to understand the context, and flexible enough!

> I also submitted the DJV project to Weblate though the website says it may take up to a month to process the submission.

Great to hear! Wait and see :) I’ll start thinking of the documentation in the meantime, and open a new pull request soon.

**Comment by darbyjohnston** [comment]
OK, it's merged. I'll take a look at some of the manual cleanup tomorrow, if it looks like it can be done in parallel I'll let you know, thanks for the offer.

Something I realized looking through the changes is that ids used in multiple places may also need to be moved to a different file; for example the id "image_controls_section_color" contains the generic word "Color" that can presumably be used in multiple places, though it is defined in djvViewApp.*.text which is specific to the viewer app. In that case the id should probably be renamed to "color" and moved to the base library djvCore.

> "Stickleback with three thorns from the inside"

Ha ha, ouch indeed!

**Comment by PiOverFour** [comment]
> OK, it's merged. I'll take a look at some of the manual cleanup tomorrow, if it looks like it can be done in parallel I'll let you know, thanks for the offer.

Awesome!

> Something I realized looking through the changes is that ids used in multiple places may also need to be moved to a different file; for example the id "image_controls_section_color" contains the generic word "Color" that can presumably be used in multiple places, though it is defined in djvViewApp.*.text which is specific to the viewer app. In that case the id should probably be renamed to "color" and moved to the base library djvCore.

This is part of the duplication problem I was referring to, and the multiple translations some strings may have.
Intuitively, it feels like "Color" is unambiguous and needs only one string*, but it may not be in all languages. For instance, another duplicate is "Medium", which may be translated to French as either "Moyen" or "Moyenne", with or without an -s suffix, depending on what it’s referring to. (French is notorious to English-speaking learners for having arbitrary grammatical genders!).
So that means we can’t assume a translation can be unique. And I thus believe we need to not only move the strings to the proper file, but create new IDs for each duplicated instance, in their respective files.

This is something I can still do.

\* I didn’t actually think Color could have multiple translations in any language, but a look at the [Japanese Wiktionary](https://ja.wiktionary.org/wiki/color#%E5%90%8D%E8%A9%9E) tells me it might: a specific color, or color as opposed to black and white. I don’t speak the language though, so I don’t know if the example is relevant.

**Comment by darbyjohnston** [comment]
> This is part of the duplication problem I was referring to, and the multiple translations some strings may have.

Ah, sorry about that, I guess I needed to see it up close to get it. :)

> Intuitively, it feels like "Color" is unambiguous and needs only one string*, but it may not be in all languages. 

That's a good point, I'll make new ids for the duplicates. I have also been adding a prefix to some of the generic ids to give them more context, especially for groups of strings that correspond to items in an enum. For example the ResourcePath enum in djvCore/Path.cpp had generic ids like "audio", "fonts", etc; I changed these to "resource_path_audio" and "resource_path_fonts" to make them more specific and keep them grouped together.

I've fixed up and checked in the strings for djvCore and the style strings in djvUI that were causing the crash earlier. If you are interested in working on some of them, maybe take a pass, see how much you have the patience to go through and I can fix up any remainders. :)

Thanks, Darby

**Comment by aberkl** [comment]
A slightly off-topic question: I found most of the DESCRIPTION to be empty:

`"description": ""`

except those:

`"description": "This is a locale label."`

Is this a value that is not used by DJV? If so I would suggest to use this as a "work in progress reminder":

`"description": "I am not too happy with this translation"`

**Comment by PiOverFour** [comment]
That value was indeed not used by DJV, but this pull request removed it entirely, as well as the `source` field which was mostly identical to the `id` field. This is to simplify manual translation and allow using Weblate in the near future.
In Weblate, strings can be marked as "Needs editing", for others to review. One can also suggest translations if unsure, as well as comment each string.

**Comment by darbyjohnston** [comment]
Just an update that I haven't heard back from Weblate about the account yet...

**Comment by PiOverFour** [comment]
Well, Weblate’s hosting page states to “please have patience”. The project is hosted in the Czech Republic, which apparently goes into quarantine today, so I guess we should indeed have patience!

**Comment by darbyjohnston** [comment]
Indeed, I just wanted to post an update that I hadn't forgotten about it, but I'm definitely in no hurry.

**Comment by PiOverFour** [comment]
Thanks :) I’m still up for writing that documentation when the system is running.
